File: kotlin/sdk/src/commonMain/kotlin/com/basecamp/sdk/services/BaseService.kt
package com.basecamp.sdk.services

import com.basecamp.sdk.*
import com.basecamp.sdk.http.BasecampHttpClient
import com.basecamp.sdk.http.currentTimeMillis
import com.basecamp.sdk.http.millisToDuration
import io.ktor.client.statement.*
import io.ktor.http.*
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Abstract base class for all Basecamp API services.
 *
 * Provides shared functionality for making API requests, handling errors,
 * and integrating with the hooks system. Generated service classes extend this.
 *
 * ```kotlin
 * class TodosService(client: AccountClient) : BaseService(client) {
 *     suspend fun list(projectId: Long, todolistId: Long): ListResult<Todo> =
 *         requestPaginated(
 *             OperationInfo("Todos", "ListTodos", "todo", false, projectId),
 *         ) {
 *             httpGet("/buckets/$projectId/todolists/$todolistId/todos.json")
 *         }
 * }
 * ```
 */
abstract class BaseService(
    private val accountClient: AccountClient,
) {
    private val http: BasecampHttpClient get() = accountClient.httpClient
    private val hooks: BasecampHooks get() = accountClient.parent.hooks
    protected val json: Json get() = http.json

    /** Maximum pages to follow as a safety cap against infinite loops. */
    private val maxPages: Int get() = accountClient.parent.config.maxPages

    /**
     * Builds the full API URL for a path relative to the account.
     * E.g., "/projects.json" -> "https://3.basecampapi.com/{accountId}/projects.json"
     */
    protected fun accountUrl(path: String): String {
        val base = accountClient.parent.config.baseUrl.trimEnd('/')
        val accountId = accountClient.accountId
        val normalizedPath = if (path.startsWith("/")) path else "/$path"
        return "$base/$accountId$normalizedPath"
    }

    /**
     * Builds a query string from key-value pairs, URL-encoding values.
     * Null values are omitted. Returns "" if no params, or "?k1=v1&k2=v2".
     */
    protected fun buildQueryString(vararg params: Pair<String, Any?>): String {
        val parts = params.mapNotNull { (key, value) ->
            value?.let { "$key=${it.toString().encodeURLParameter()}" }
        }
        return if (parts.isEmpty()) "" else "?" + parts.joinToString("&")
    }

    /**
     * Executes a GET request for the given account-relative path.
     */
    protected suspend fun httpGet(path: String, operationName: String? = null): HttpResponse =
        http.requestWithRetry(HttpMethod.Get, accountUrl(path), operationName = operationName)

    /**
     * Executes a POST request with a JSON body.
     */
    protected suspend fun httpPost(path: String, body: String? = null, operationName: String? = null): HttpResponse =
        http.requestWithRetry(HttpMethod.Post, accountUrl(path), body, operationName = operationName)

    /**
     * Executes a PUT request with a JSON body.
     */
    protected suspend fun httpPut(path: String, body: String? = null, operationName: String? = null): HttpResponse =
        http.requestWithRetry(HttpMethod.Put, accountUrl(path), body, operationName = operationName)

    /**
     * Executes a DELETE request.
     */
    protected suspend fun httpDelete(path: String, operationName: String? = null): HttpResponse =
        http.requestWithRetry(HttpMethod.Delete, accountUrl(path), operationName = operationName)

    /**
     * Executes a POST request with binary body data.
     */
    protected suspend fun httpPostBinary(path: String, data: ByteArray, contentType: String): HttpResponse =
        http.requestBinaryWithRetry(HttpMethod.Post, accountUrl(path), data, contentType)

    /**
     * Executes an API request with error handling and hooks integration.
     *
     * @param info Operation metadata for hooks.
     * @param fn The suspend function that performs the actual HTTP call.
     * @param parse Deserializes the response body string into the result type.
     * @return The parsed response.
     */
    protected suspend fun <T> request(
        info: OperationInfo,
        fn: suspend () -> HttpResponse,
        parse: (String) -> T,
    ): T {
        val startTime = currentTimeMillis()

        hooks.safeOnOperationStart(info)

        try {
            val response = fn()
            val duration = (currentTimeMillis() - startTime).millisToDuration()

            if (!response.status.isSuccess()) {
                val error = errorFromResponse(response)
                hooks.safeOnOperationEnd(info, OperationResult(duration, error))
                throw error
            }

            // 204 No Content
            if (response.status.value == 204) {
                hooks.safeOnOperationEnd(info, OperationResult(duration))
                @Suppress("UNCHECKED_CAST")
                return Unit as T
            }

            val bodyText = response.bodyAsText()
            val result = parse(bodyText)
            hooks.safeOnOperationEnd(info, OperationResult(duration))
            return result
        } catch (e: BasecampException) {
            val duration = (currentTimeMillis() - startTime).millisToDuration()
            hooks.safeOnOperationEnd(info, OperationResult(duration, e))
            throw e
        } catch (e: Exception) {
            val duration = (currentTimeMillis() - startTime).millisToDuration()
            hooks.safeOnOperationEnd(info, OperationResult(duration, e))
            throw e
        }
    }

    /**
     * Executes a paginated API request, automatically following Link headers.
     *
     * Returns a [ListResult] with all items across pages, plus [ListMeta]
     * with `totalCount` and `truncated` information.
     *
     * @param info Operation metadata for hooks.
     * @param options Pagination control (maxItems).
     * @param fn The suspend function that performs the initial HTTP call.
     * @param parseItems Parses a page's response body into a list of items.
     */
    protected suspend fun <T> requestPaginated(
        info: OperationInfo,
        options: PaginationOptions? = null,
        fn: suspend () -> HttpResponse,
        parseItems: (String) -> List<T>,
    ): ListResult<T> {
        val startTime = currentTimeMillis()
        val maxItems = options?.maxItems

        hooks.safeOnOperationStart(info)

        try {
            val response = fn()

            if (!response.status.isSuccess()) {
                val error = errorFromResponse(response)
                val duration = (currentTimeMillis() - startTime).millisToDuration()
                hooks.safeOnOperationEnd(info, OperationResult(duration, error))
                throw error
            }

            val bodyText = response.bodyAsText()
            val firstPageItems = parseItems(bodyText)
            val totalCount = parseTotalCount(response.headers.toMap())

            // Check if maxItems is satisfied by the first page
            if (maxItems != null && maxItems > 0 && firstPageItems.size >= maxItems) {
                val hasMore = firstPageItems.size > maxItems
                    || parseNextLink(response.headers["Link"]) != null
                val duration = (currentTimeMillis() - startTime).millisToDuration()
                val result = ListResult(firstPageItems.take(maxItems), ListMeta(totalCount, hasMore))
                hooks.safeOnOperationEnd(info, OperationResult(duration))
                return result
            }

            // Follow pagination
            val allItems = firstPageItems.toMutableList()
            var currentResponse = response
            val initialUrl = response.request.url.toString()

            for (page in 1 until maxPages) {
                val rawNextUrl = parseNextLink(currentResponse.headers["Link"]) ?: break
                val nextUrl = resolveUrl(currentResponse.request.url.toString(), rawNextUrl)

                // Validate same-origin to prevent SSRF / token leakage
                if (!isSameOrigin(nextUrl, initialUrl)) {
                    throw BasecampException.Api(
                        "Pagination Link header points to different origin: $nextUrl",
                        httpStatus = 0,
                    )
                }

                currentResponse = http.requestWithRetry(HttpMethod.Get, nextUrl)

                if (!currentResponse.status.isSuccess()) {
                    throw errorFromResponse(currentResponse)
                }

                val pageBody = currentResponse.bodyAsText()
                val pageItems = parseItems(pageBody)
                allItems.addAll(pageItems)

                // Check maxItems cap
                if (maxItems != null && maxItems > 0 && allItems.size >= maxItems) {
                    val duration = (currentTimeMillis() - startTime).millisToDuration()
                    val result = ListResult(allItems.take(maxItems), ListMeta(totalCount, truncated = true))
                    hooks.safeOnOperationEnd(info, OperationResult(duration))
                    return result
                }
            }

            val hasMore = parseNextLink(currentResponse.headers["Link"]) != null
            val duration = (currentTimeMillis() - startTime).millisToDuration()
            val result = ListResult(allItems, ListMeta(totalCount, hasMore))
            hooks.safeOnOperationEnd(info, OperationResult(duration))
            return result
        } catch (e: BasecampException) {
            val duration = (currentTimeMillis() - startTime).millisToDuration()
            hooks.safeOnOperationEnd(info, OperationResult(duration, e))
            throw e
        } catch (e: Exception) {
            val duration = (currentTimeMillis() - startTime).millisToDuration()
            hooks.safeOnOperationEnd(info, OperationResult(duration, e))
            throw e
        }
    }

    /**
     * Executes a paginated request for wrapped responses, returning both the raw
     * first page body (for wrapper field decoding) and the paginated items.
     *
     * @param info Operation metadata for hooks.
     * @param options Pagination control (maxItems).
     * @param fn The suspend function that performs the initial HTTP call.
     * @param parseItems Parses a page's response body into a list of items.
     * @return A [Pair] of the first page's raw body and the [ListResult] of all items.
     */
    protected suspend fun <T> requestPaginatedWrapped(
        info: OperationInfo,
        options: PaginationOptions? = null,
        fn: suspend () -> HttpResponse,
        parseItems: (String) -> List<T>,
    ): Pair<String, ListResult<T>> {
        val startTime = currentTimeMillis()
        val maxItems = options?.maxItems

        hooks.safeOnOperationStart(info)

        try {
            val response = fn()

            if (!response.status.isSuccess()) {
                val error = errorFromResponse(response)
                val duration = (currentTimeMillis() - startTime).millisToDuration()
                hooks.safeOnOperationEnd(info, OperationResult(duration, error))
                throw error
            }

            val firstPageBody = response.bodyAsText()
            val firstPageItems = parseItems(firstPageBody)
            val totalCount = parseTotalCount(response.headers.toMap())

            // Check if maxItems is satisfied by the first page
            if (maxItems != null && maxItems > 0 && firstPageItems.size >= maxItems) {
                val hasMore = firstPageItems.size > maxItems
                    || parseNextLink(response.headers["Link"]) != null
                val duration = (currentTimeMillis() - startTime).millisToDuration()
                val result = ListResult(firstPageItems.take(maxItems), ListMeta(totalCount, hasMore))
                hooks.safeOnOperationEnd(info, OperationResult(duration))
                return Pair(firstPageBody, result)
            }

            // Follow pagination
            val allItems = firstPageItems.toMutableList()
            var currentResponse = response
            val initialUrl = response.request.url.toString()

            for (page in 1 until maxPages) {
                val rawNextUrl = parseNextLink(currentResponse.headers["Link"]) ?: break
                val nextUrl = resolveUrl(currentResponse.request.url.toString(), rawNextUrl)

                if (!isSameOrigin(nextUrl, initialUrl)) {
                    throw BasecampException.Api(
                        "Pagination Link header points to different origin: $nextUrl",
                        httpStatus = 0,
                    )
                }

                currentResponse = http.requestWithRetry(HttpMethod.Get, nextUrl)

                if (!currentResponse.status.isSuccess()) {
                    throw errorFromResponse(currentResponse)
                }

                val pageBody = currentResponse.bodyAsText()
                val pageItems = parseItems(pageBody)
                allItems.addAll(pageItems)

                if (maxItems != null && maxItems > 0 && allItems.size >= maxItems) {
                    val duration = (currentTimeMillis() - startTime).millisToDuration()
                    val result = ListResult(allItems.take(maxItems), ListMeta(totalCount, truncated = true))
                    hooks.safeOnOperationEnd(info, OperationResult(duration))
                    return Pair(firstPageBody, result)
                }
            }

            val hasMore = parseNextLink(currentResponse.headers["Link"]) != null
            val duration = (currentTimeMillis() - startTime).millisToDuration()
            val result = ListResult(allItems, ListMeta(totalCount, hasMore))
            hooks.safeOnOperationEnd(info, OperationResult(duration))
            return Pair(firstPageBody, result)
        } catch (e: BasecampException) {
            val duration = (currentTimeMillis() - startTime).millisToDuration()
            hooks.safeOnOperationEnd(info, OperationResult(duration, e))
            throw e
        } catch (e: Exception) {
            val duration = (currentTimeMillis() - startTime).millisToDuration()
            hooks.safeOnOperationEnd(info, OperationResult(duration, e))
            throw e
        }
    }

    /**
     * Streaming paginated request that emits items as each page arrives.
     *
     * Unlike [requestPaginated] which eagerly loads all pages, this returns
     * a cold [Flow] that fetches pages lazily as the collector consumes items.
     * Useful for processing large datasets without loading everything into memory.
     *
     * ```kotlin
     * account.todos.listAsFlow(projectId, todolistId)
     *     .collect { todo -> println(todo.content) }
     * ```
     *
     * @param info Operation metadata for hooks.
     * @param fn The suspend function that performs the initial HTTP call.
     * @param parseItems Parses a page's response body into a list of items.
     */
    protected fun <T> requestPaginatedAsFlow(
        info: OperationInfo,
        fn: suspend () -> HttpResponse,
        parseItems: (String) -> List<T>,
    ): Flow<T> = flow {
        val startTime = currentTimeMillis()
        hooks.safeOnOperationStart(info)

        try {
            var currentResponse = fn()

            if (!currentResponse.status.isSuccess()) {
                throw errorFromResponse(currentResponse)
            }

            val bodyText = currentResponse.bodyAsText()
            val firstPageItems = parseItems(bodyText)
            for (item in firstPageItems) emit(item)

            val initialUrl = currentResponse.request.url.toString()

            for (page in 1 until maxPages) {
                val rawNextUrl = parseNextLink(currentResponse.headers["Link"]) ?: break
                val nextUrl = resolveUrl(currentResponse.request.url.toString(), rawNextUrl)

                if (!isSameOrigin(nextUrl, initialUrl)) {
                    throw BasecampException.Api(
                        "Pagination Link header points to different origin: $nextUrl",
                        httpStatus = 0,
                    )
                }

                currentResponse = http.requestWithRetry(HttpMethod.Get, nextUrl)

                if (!currentResponse.status.isSuccess()) {
                    throw errorFromResponse(currentResponse)
                }

                val pageBody = currentResponse.bodyAsText()
                val pageItems = parseItems(pageBody)
                for (item in pageItems) emit(item)
            }

            val duration = (currentTimeMillis() - startTime).millisToDuration()
            hooks.safeOnOperationEnd(info, OperationResult(duration))
        } catch (e: Exception) {
            val duration = (currentTimeMillis() - startTime).millisToDuration()
            hooks.safeOnOperationEnd(info, OperationResult(duration, e))
            throw e
        }
    }

    /** Converts an HTTP error response to a [BasecampException]. */
    private suspend fun errorFromResponse(response: HttpResponse): BasecampException {
        val status = response.status.value
        val requestId = response.headers["X-Request-Id"]
        val retryAfter = parseRetryAfter(response.headers["Retry-After"])

        var message: String = response.status.description.ifEmpty { "Request failed" }
        var hint: String? = null

        try {
            val bodyText = response.bodyAsText()
            if (bodyText.isNotBlank()) {
                val jsonBody = json.parseToJsonElement(bodyText)
                if (jsonBody is JsonObject) {
                    jsonBody["error"]?.jsonPrimitive?.content?.let {
                        message = BasecampException.truncateMessage(it)
                    }
                    jsonBody["error_description"]?.jsonPrimitive?.content?.let {
                        hint = BasecampException.truncateMessage(it)
                    }
                }
            }
        } catch (_: Exception) {
            // Body is not JSON or empty — use status text
        }

        return BasecampException.fromHttpStatus(status, message, hint, requestId, retryAfter)
    }

    companion object {
        /** Resolve a potentially relative URL against a base URL. */
        internal fun resolveUrl(base: String, relative: String): String {
            // If it's already absolute, return as-is
            if (relative.startsWith("http://") || relative.startsWith("https://")) {
                return relative
            }
            // Extract origin from base
            val schemeEnd = base.indexOf("://")
            if (schemeEnd < 0) return relative
            val afterScheme = schemeEnd + 3
            val pathStart = base.indexOf('/', afterScheme)
            val origin = if (pathStart < 0) base else base.substring(0, pathStart)
            val normalizedPath = if (relative.startsWith("/")) relative else "/$relative"
            return "$origin$normalizedPath"
        }
    }
}

/** Safely invoke onOperationStart, catching hook exceptions. */
private fun BasecampHooks.safeOnOperationStart(info: OperationInfo) {
    runCatching { onOperationStart(info) }
}

/** Safely invoke onOperationEnd, catching hook exceptions. */
private fun BasecampHooks.safeOnOperationEnd(info: OperationInfo, result: OperationResult) {
    runCatching { onOperationEnd(info, result) }
}

/** Convert Ktor headers to a simple map for pagination utilities. */
private fun io.ktor.http.Headers.toMap(): Map<String, List<String>> {
    val result = mutableMapOf<String, List<String>>()
    forEach { key, values -> result[key] = values }
    return result
}

File: swift/Sources/Basecamp/Services/BaseService.swift
import Foundation

/// Base class for all Basecamp API services.
///
/// Provides shared functionality for making API requests, handling errors,
/// automatic pagination via Link headers, and hooks integration.
///
/// Generated services inherit from this class. The `request()`, `requestVoid()`,
/// and `requestPaginated()` methods handle the full operation lifecycle.
open class BaseService: @unchecked Sendable {
    /// The account client this service is bound to.
    public let accountClient: AccountClient

    /// Creates a service bound to an account client.
    public init(accountClient: AccountClient) {
        self.accountClient = accountClient
    }

    // MARK: - Request Methods

    /// Executes an API request and decodes the response.
    ///
    /// Handles error mapping, hooks lifecycle (operation start/end),
    /// and response deserialization.
    ///
    /// - Parameters:
    ///   - info: Operation metadata for hooks.
    ///   - method: HTTP method.
    ///   - path: URL path relative to the account base (e.g., "/projects.json").
    ///   - body: Optional encodable request body.
    ///   - retryConfig: Optional per-operation retry configuration.
    /// - Returns: The decoded response.
    public func request<T: Decodable & Sendable>(
        _ info: OperationInfo,
        method: String,
        path: String,
        body: (any Encodable & Sendable)? = nil,
        contentType: String? = nil,
        retryConfig: RetryConfig? = nil
    ) async throws -> T {
        let startTime = CFAbsoluteTimeGetCurrent()

        safeInvokeHooks { $0.onOperationStart(info) }

        do {
            let bodyData: Data?
            if contentType != nil {
                bodyData = body as? Data
            } else {
                bodyData = try body.map { try Self.encoder.encode($0) }
            }
            let url = accountClient.baseURL + path

            let (data, response) = try await accountClient.httpClient.performRequest(
                method: method, url: url, body: bodyData, contentType: contentType, retryConfig: retryConfig
            )

            let durationMs = millisSince(startTime)

            guard response.statusCode >= 200 && response.statusCode < 300 else {
                throw BasecampError.fromHTTPResponse(
                    status: response.statusCode, data: data,
                    headers: response.allHeaderFields as? [String: String] ?? [:],
                    requestId: response.value(forHTTPHeaderField: "X-Request-Id")
                )
            }

            let decoded = try Self.decoder.decode(T.self, from: data)
            safeInvokeHooks { $0.onOperationEnd(info, result: OperationResult(durationMs: durationMs)) }
            return decoded
        } catch {
            let durationMs = millisSince(startTime)
            safeInvokeHooks { $0.onOperationEnd(info, result: OperationResult(durationMs: durationMs, error: error)) }
            throw error
        }
    }

    /// Executes an API request that returns no response body (e.g., DELETE, PUT status changes).
    ///
    /// - Parameters:
    ///   - info: Operation metadata for hooks.
    ///   - method: HTTP method.
    ///   - path: URL path relative to the account base.
    ///   - body: Optional encodable request body.
    ///   - retryConfig: Optional per-operation retry configuration.
    public func requestVoid(
        _ info: OperationInfo,
        method: String,
        path: String,
        body: (any Encodable & Sendable)? = nil,
        contentType: String? = nil,
        retryConfig: RetryConfig? = nil
    ) async throws {
        let startTime = CFAbsoluteTimeGetCurrent()

        safeInvokeHooks { $0.onOperationStart(info) }

        do {
            let bodyData: Data?
            if contentType != nil {
                bodyData = body as? Data
            } else {
                bodyData = try body.map { try Self.encoder.encode($0) }
            }
            let url = accountClient.baseURL + path

            let (data, response) = try await accountClient.httpClient.performRequest(
                method: method, url: url, body: bodyData, contentType: contentType, retryConfig: retryConfig
            )

            let durationMs = millisSince(startTime)

            guard response.statusCode >= 200 && response.statusCode < 300 else {
                throw BasecampError.fromHTTPResponse(
                    status: response.statusCode, data: data,
                    headers: response.allHeaderFields as? [String: String] ?? [:],
                    requestId: response.value(forHTTPHeaderField: "X-Request-Id")
                )
            }

            safeInvokeHooks { $0.onOperationEnd(info, result: OperationResult(durationMs: durationMs)) }
        } catch {
            let durationMs = millisSince(startTime)
            safeInvokeHooks { $0.onOperationEnd(info, result: OperationResult(durationMs: durationMs, error: error)) }
            throw error
        }
    }

    /// Executes a paginated API request, automatically following Link headers.
    ///
    /// Returns a `ListResult<T>` conforming to `RandomAccessCollection`.
    ///
    /// - Parameters:
    ///   - info: Operation metadata for hooks.
    ///   - path: URL path relative to the account base.
    ///   - queryItems: Optional query parameters.
    ///   - paginationOpts: Optional pagination control.
    ///   - retryConfig: Optional per-operation retry configuration.
    /// - Returns: A `ListResult` containing all items across pages.
    public func requestPaginated<T: Decodable & Sendable>(
        _ info: OperationInfo,
        path: String,
        queryItems: [URLQueryItem]? = nil,
        paginationOpts: PaginationOptions? = nil,
        retryConfig: RetryConfig? = nil
    ) async throws -> ListResult<T> {
        let startTime = CFAbsoluteTimeGetCurrent()

        safeInvokeHooks { $0.onOperationStart(info) }

        do {
            var urlString = accountClient.baseURL + path
            if let queryItems, !queryItems.isEmpty {
                var components = URLComponents(string: urlString)
                components?.queryItems = queryItems
                urlString = components?.string ?? urlString
            }

            let (data, response) = try await accountClient.httpClient.performRequest(
                method: "GET", url: urlString, retryConfig: retryConfig
            )

            guard response.statusCode >= 200 && response.statusCode < 300 else {
                throw BasecampError.fromHTTPResponse(
                    status: response.statusCode, data: data,
                    headers: response.allHeaderFields as? [String: String] ?? [:],
                    requestId: response.value(forHTTPHeaderField: "X-Request-Id")
                )
            }

            let firstPageItems = try Self.decoder.decode([T].self, from: data)
            let totalCount = parseTotalCount(response)
            let maxItems = paginationOpts?.maxItems

            // If maxItems is set and first page satisfies it, return early
            if let maxItems, maxItems > 0, firstPageItems.count >= maxItems {
                let hasMore = firstPageItems.count > maxItems
                    || parseNextLink(response.value(forHTTPHeaderField: "Link")) != nil
                let durationMs = millisSince(startTime)
                safeInvokeHooks { $0.onOperationEnd(info, result: OperationResult(durationMs: durationMs)) }
                return ListResult(
                    Array(firstPageItems.prefix(maxItems)),
                    meta: ListMeta(totalCount: totalCount, truncated: hasMore)
                )
            }

            // Follow pagination
            let (allItems, truncated) = try await followPagination(
                initialURL: urlString,
                initialResponse: response,
                firstPageItems: firstPageItems,
                maxItems: maxItems
            )

            let durationMs = millisSince(startTime)
            safeInvokeHooks { $0.onOperationEnd(info, result: OperationResult(durationMs: durationMs)) }
            return ListResult(allItems, meta: ListMeta(totalCount: totalCount, truncated: truncated))
        } catch {
            let durationMs = millisSince(startTime)
            safeInvokeHooks { $0.onOperationEnd(info, result: OperationResult(durationMs: durationMs, error: error)) }
            throw error
        }
    }

    /// Executes a paginated request for wrapped responses, returning both the
    /// raw first page data (for wrapper field decoding) and the paginated items.
    ///
    /// Each page returns `{ key: [items], ... }` — items are extracted from the given key.
    /// The first page's raw data is returned so callers can decode wrapper fields like `person`.
    ///
    /// - Parameters:
    ///   - info: Operation metadata for hooks.
    ///   - path: URL path relative to the account base.
    ///   - itemsKey: The JSON key containing the array of items in each response.
    ///   - queryItems: Optional query parameters.
    ///   - paginationOpts: Optional pagination control.
    ///   - retryConfig: Optional per-operation retry configuration.
    /// - Returns: A tuple of the first page's raw data and a `ListResult` of all items.
    public func requestPaginatedWrapped<T: Decodable & Sendable>(
        _ info: OperationInfo,
        path: String,
        itemsKey: String,
        queryItems: [URLQueryItem]? = nil,
        paginationOpts: PaginationOptions? = nil,
        retryConfig: RetryConfig? = nil
    ) async throws -> (Data, ListResult<T>) {
        let startTime = CFAbsoluteTimeGetCurrent()

        safeInvokeHooks { $0.onOperationStart(info) }

        do {
            var urlString = accountClient.baseURL + path
            if let queryItems, !queryItems.isEmpty {
                var components = URLComponents(string: urlString)
                components?.queryItems = queryItems
                urlString = components?.string ?? urlString
            }

            let (data, response) = try await accountClient.httpClient.performRequest(
                method: "GET", url: urlString, retryConfig: retryConfig
            )

            guard response.statusCode >= 200 && response.statusCode < 300 else {
                throw BasecampError.fromHTTPResponse(
                    status: response.statusCode, data: data,
                    headers: response.allHeaderFields as? [String: String] ?? [:],
                    requestId: response.value(forHTTPHeaderField: "X-Request-Id")
                )
            }

            let firstPageData = data
            let firstPageItems: [T] = try Self.decodeWrappedItems(data: data, key: itemsKey)
            let totalCount = parseTotalCount(response)
            let maxItems = paginationOpts?.maxItems

            // If maxItems is set and first page satisfies it, return early
            if let maxItems, maxItems > 0, firstPageItems.count >= maxItems {
                let hasMore = firstPageItems.count > maxItems
                    || parseNextLink(response.value(forHTTPHeaderField: "Link")) != nil
                let durationMs = millisSince(startTime)
                safeInvokeHooks { $0.onOperationEnd(info, result: OperationResult(durationMs: durationMs)) }
                return (firstPageData, ListResult(
                    Array(firstPageItems.prefix(maxItems)),
                    meta: ListMeta(totalCount: totalCount, truncated: hasMore)
                ))
            }

            // Follow pagination
            let (allItems, truncated) = try await followWrappedPagination(
                initialURL: urlString,
                initialResponse: response,
                firstPageItems: firstPageItems,
                itemsKey: itemsKey,
                maxItems: maxItems
            )

            let durationMs = millisSince(startTime)
            safeInvokeHooks { $0.onOperationEnd(info, result: OperationResult(durationMs: durationMs)) }
            return (firstPageData, ListResult(allItems, meta: ListMeta(totalCount: totalCount, truncated: truncated)))
        } catch {
            let durationMs = millisSince(startTime)
            safeInvokeHooks { $0.onOperationEnd(info, result: OperationResult(durationMs: durationMs, error: error)) }
            throw error
        }
    }

    // MARK: - Pagination

    private func followPagination<T: Decodable>(
        initialURL: String,
        initialResponse: HTTPURLResponse,
        firstPageItems: [T],
        maxItems: Int?
    ) async throws -> (items: [T], truncated: Bool) {
        var allItems = firstPageItems
        var response = initialResponse
        let maxPages = accountClient.maxPages

        for _ in 1..<maxPages {
            guard let rawNextURL = parseNextLink(response.value(forHTTPHeaderField: "Link")) else {
                break
            }

            let nextURL = resolveURL(base: response.url?.absoluteString ?? initialURL, target: rawNextURL)

            // Validate same-origin to prevent SSRF / token leakage
            guard isSameOrigin(nextURL, initialURL) else {
                throw BasecampError.api(
                    message: "Pagination Link header points to different origin: \(nextURL)",
                    httpStatus: nil, hint: nil, requestId: nil
                )
            }

            let (data, nextResponse) = try await accountClient.httpClient.fetchPage(url: nextURL)

            guard nextResponse.statusCode >= 200 && nextResponse.statusCode < 300 else {
                throw BasecampError.fromHTTPResponse(
                    status: nextResponse.statusCode, data: data,
                    headers: nextResponse.allHeaderFields as? [String: String] ?? [:],
                    requestId: nextResponse.value(forHTTPHeaderField: "X-Request-Id")
                )
            }

            let pageItems = try Self.decoder.decode([T].self, from: data)
            allItems.append(contentsOf: pageItems)

            // Check maxItems cap
            if let maxItems, maxItems > 0, allItems.count >= maxItems {
                return (Array(allItems.prefix(maxItems)), true)
            }

            response = nextResponse
        }

        // If we hit the page cap and there's still a next link, results are truncated
        let hasMore = parseNextLink(response.value(forHTTPHeaderField: "Link")) != nil
        return (allItems, hasMore)
    }

    private func followWrappedPagination<T: Decodable>(
        initialURL: String,
        initialResponse: HTTPURLResponse,
        firstPageItems: [T],
        itemsKey: String,
        maxItems: Int?
    ) async throws -> (items: [T], truncated: Bool) {
        var allItems = firstPageItems
        var response = initialResponse
        let maxPages = accountClient.maxPages

        for _ in 1..<maxPages {
            guard let rawNextURL = parseNextLink(response.value(forHTTPHeaderField: "Link")) else {
                break
            }

            let nextURL = resolveURL(base: response.url?.absoluteString ?? initialURL, target: rawNextURL)

            // Validate same-origin to prevent SSRF / token leakage
            guard isSameOrigin(nextURL, initialURL) else {
                throw BasecampError.api(
                    message: "Pagination Link header points to different origin: \(nextURL)",
                    httpStatus: nil, hint: nil, requestId: nil
                )
            }

            let (data, nextResponse) = try await accountClient.httpClient.fetchPage(url: nextURL)

            guard nextResponse.statusCode >= 200 && nextResponse.statusCode < 300 else {
                throw BasecampError.fromHTTPResponse(
                    status: nextResponse.statusCode, data: data,
                    headers: nextResponse.allHeaderFields as? [String: String] ?? [:],
                    requestId: nextResponse.value(forHTTPHeaderField: "X-Request-Id")
                )
            }

            let pageItems: [T] = try Self.decodeWrappedItems(data: data, key: itemsKey)
            allItems.append(contentsOf: pageItems)

            // Check maxItems cap
            if let maxItems, maxItems > 0, allItems.count >= maxItems {
                return (Array(allItems.prefix(maxItems)), true)
            }

            response = nextResponse
        }

        // If we hit the page cap and there's still a next link, results are truncated
        let hasMore = parseNextLink(response.value(forHTTPHeaderField: "Link")) != nil
        return (allItems, hasMore)
    }

    /// Decodes items from a wrapped JSON response by extracting the array at the given key.
    private static func decodeWrappedItems<T: Decodable>(data: Data, key: String) throws -> [T] {
        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]
        guard let itemsArray = json[key] else {
            return []
        }
        let itemsData = try JSONSerialization.data(withJSONObject: itemsArray)
        return try decoder.decode([T].self, from: itemsData)
    }

    // MARK: - Shared Coders

    /// Shared JSON decoder configured for the Basecamp API.
    public static let decoder: JSONDecoder = {
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        return decoder
    }()

    /// Shared JSON encoder configured for the Basecamp API.
    public static let encoder: JSONEncoder = {
        let encoder = JSONEncoder()
        encoder.keyEncodingStrategy = .convertToSnakeCase
        return encoder
    }()

    // MARK: - Helpers

    private func millisSince(_ startTime: CFAbsoluteTime) -> Int {
        Int((CFAbsoluteTimeGetCurrent() - startTime) * 1000)
    }

    private func safeInvokeHooks(_ invoke: (any BasecampHooks) -> Void) {
        invoke(accountClient.hooks)
    }
}
